File: backend/src/main/java/com/nexustasks/task/entity/Task.java
package com.nexustasks.task.entity;

import com.nexustasks.category.entity.Category;
import com.nexustasks.common.entity.BaseEntity;
import com.nexustasks.user.entity.User;
import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;

@Entity
@Table(name = "tasks")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Task extends BaseEntity {

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private TaskStatus status = TaskStatus.TODO;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private TaskPriority priority = TaskPriority.MEDIUM;

    @Column(name = "due_date")
    private Instant dueDate;

    @Column(name = "completed_at")
    private Instant completedAt;

    @Column(nullable = false)
    @Builder.Default
    private boolean archived = false;

    @Column(name = "archived_at")
    private Instant archivedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    /**
     * Callback JPA pour maintenir la cohérence des dates d'état.
     */
    @PreUpdate
    @PrePersist
    public void updateStateTimestamps() {
        // Gérer completedAt
        if (this.status == TaskStatus.COMPLETED && this.completedAt == null) {
            this.completedAt = Instant.now();
        } else if (this.status != TaskStatus.COMPLETED) {
            this.completedAt = null; // Reset si on repasse à TODO/IN_PROGRESS
        }

        // Gérer archivedAt (booléen métier 'archived')
        if (this.archived && this.archivedAt == null) {
            this.archivedAt = Instant.now();
        } else if (!this.archived) {
            this.archivedAt = null; // Reset si désarchivé
        }
    }

}

File: mobile/lib/features/tasks/domain/models/task.dart
class TaskCategory {
  const TaskCategory({
    required this.publicId,
    required this.name,
    required this.color,
  });

  final String publicId;
  final String name;
  final String color;

  factory TaskCategory.fromJson(Map<String, dynamic> json) => TaskCategory(
    publicId: json['publicId'] as String,
    name: json['name'] as String,
    color: (json['color'] as String?) ?? '#6b7280',
  );
}

class Task {
  const Task({
    required this.publicId,
    required this.title,
    required this.status,
    required this.priority,
    this.description,
    this.dueDate,
    this.completedAt,
    required this.archived,
    required this.createdAt,
    required this.updatedAt,
    this.category,
  });

  final String publicId;
  final String title;
  final String status;
  final String priority;
  final String? description;
  final DateTime? dueDate;
  final DateTime? completedAt;
  final bool archived;
  final DateTime createdAt;
  final DateTime updatedAt;
  final TaskCategory? category;

  factory Task.fromJson(Map<String, dynamic> json) => Task(
    publicId: json['publicId'] as String,
    title: json['title'] as String,
    status: json['status'] as String,
    priority: json['priority'] as String,
    description: json['description'] as String?,
    dueDate: json['dueDate'] != null
        ? DateTime.parse(json['dueDate'] as String)
        : null,
    completedAt: json['completedAt'] != null
        ? DateTime.parse(json['completedAt'] as String)
        : null,
    archived: (json['archived'] as bool?) ?? false,
    createdAt: DateTime.parse(json['createdAt'] as String),
    updatedAt: DateTime.parse(json['updatedAt'] as String),
    category: json['category'] != null
        ? TaskCategory.fromJson(json['category'] as Map<String, dynamic>)
        : null,
  );
}
